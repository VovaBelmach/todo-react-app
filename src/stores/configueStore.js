import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todosSlice";

const loadStateFromLocalStorage = () => {
  const storedTodos = localStorage.getItem("todos");
  if (storedTodos) {
    return {
      todos: {
        todos: JSON.parse(storedTodos),
        filterValue: "All",
      },
    };
  }
  return undefined;
};

const persistedState = loadStateFromLocalStorage();

export default configureStore({
  reducer: {
    todos: todoReducer,
  },
  preloadedState: persistedState
});