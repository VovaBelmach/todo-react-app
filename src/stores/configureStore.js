import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todosSlice";
import { loadTodosStateFromLocalStorage } from "./helpers/todoLocalStorage.helper";
import { localStorageMiddleware } from "./middleware/localStorageRepositoryMiddleware"

export default configureStore({
  reducer: {
    todos: todoReducer,
  },
  preloadedState: loadTodosStateFromLocalStorage(),
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(localStorageMiddleware),
});
