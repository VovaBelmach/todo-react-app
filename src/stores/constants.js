export const TODOS_LOCAL_STORAGE_NAME = "todos";