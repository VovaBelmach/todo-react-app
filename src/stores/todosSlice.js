import { createSlice } from "@reduxjs/toolkit";
import { v4 as uuid } from "uuid";
import { saveItemsToLocalStorage } from "../repositories/localStorageRepository";
import {
  TODOS_LOCAL_STORAGE_NAME,
  TODO_FILTER_ALL_BUTTON_NAME,
  TODO_FILTER_ACTIVE_BUTTON_NAME,
  TODO_FILTER_COMPLITED_BUTTON_NAME
} from "../constants";

const initialState = {
  todos: [],
  filterValue: TODO_FILTER_ALL_BUTTON_NAME,
};

export const todosSlice = createSlice({
  name: TODOS_LOCAL_STORAGE_NAME,
  initialState: initialState,
  reducers: {
    addTodo: (state, action) => {
      const newTodo = {
        id: uuid(),
        description: action.payload.description,
        isCompleted: false,
      };
      state.todos.push(newTodo);
      saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
    },
    deleteTodo: (state, action) => {
      const { id } = action.payload;
      state.todos = state.todos.filter((todo) => todo.id !== id);
      saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
    },
    completeTodo: (state, action) => {
      const index = state.todos.findIndex(
        (item) => item.id === action.payload.id
      );
      if (index !== -1) {
        state.todos[index].isCompleted = !state.todos[index].isCompleted;
        saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
      }
    },
    setFilter: (state, action) => {
      state.filterValue = action.payload;
    },
    reorderTodos: (state, action) => {
      const { draggedIndex, droppedIndex } = action.payload;
      const draggedTodo = state.todos[draggedIndex];
      const newTodos = [...state.todos];
      newTodos.splice(draggedIndex, 1);
      newTodos.splice(droppedIndex, 0, draggedTodo);
      state.todos = newTodos;
      saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
    },
  },
});

export const { addTodo, deleteTodo, completeTodo, setFilter, reorderTodos } =
  todosSlice.actions;

export default todosSlice.reducer;

export const filteredTodosSelector = (state) => {
  const { todos, filterValue } = state.todos;
  switch (filterValue) {
    case TODO_FILTER_ACTIVE_BUTTON_NAME:
      return todos.filter((todo) => todo.isCompleted !== true);
    case TODO_FILTER_COMPLITED_BUTTON_NAME:
      return todos.filter((todo) => todo.isCompleted === true);
    default:
      return todos;
  }
};
