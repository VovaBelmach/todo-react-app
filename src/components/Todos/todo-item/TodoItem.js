import { useState } from "react";
import { useDispatch } from "react-redux";
import { deleteTodo, checkTodo } from "../../../stores/todosSlice";
import Checkbox from "../../UI/Checkbox/Checkbox";
import Button from "../../UI/Button/Button";
import styles from "./TodoItem.module.css";

const TodoItem = (props) => {
  const dispatch = useDispatch();
  const [isCompleted, setCompletition] = useState(props.isCompleted);

  const onDeleteHandler = (event) => {
    event.stopPropagation();
    dispatch(
      deleteTodo({
        id: props.id,
      })
    );
  };

  const onCheckHandler = () => {
    dispatch(
      checkTodo({
        id: props.id,
      })
    );
    setCompletition(!isCompleted);
  };

  return (
    <li
      key={props.id}
      draggable
      onDragStart={(event) => props.onDragStartHandle(event, props.index)}
      onDragOver={props.onDragOverHandle}
      onDrop={(event) => props.onDropHandle(event, props.index)}
      onDragEnd={props.onDragEndHandle}
      onClick={onCheckHandler}
      style={{
        opacity: props.draggedIndex === props.index ? 0.5 : 1,
        cursor: "move",
      }}
    >
      <Checkbox
        id={props.id}
        name={props.id}
        isCompleted={isCompleted}
        isDisabled={false}
      />
      <label>{props.description}</label>
      <Button onClick={onDeleteHandler} className={`${styles["delete-btn"]}`}>
        X
      </Button>
    </li>
  );
};

export default TodoItem;
