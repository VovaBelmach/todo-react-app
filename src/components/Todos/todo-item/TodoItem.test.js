import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import TodoItem from "./TodoItem";
import { completeTodo, deleteTodo } from "../../../stores/todosSlice";

const mockStore = configureMockStore();
const store = mockStore({});

const props = {
  id: "1",
  index: 0,
  isCompleted: false,
  description: "Test Todo",
  onDragStartHandle: jest.fn(),
  onDragOverHandle: jest.fn(),
  onDropHandle: jest.fn(),
  onDragEndHandle: jest.fn(),
};

describe("TodoItem component", () => {
  it("should render the description and a checkbox", () => {
    render(
      <Provider store={store}>
        <TodoItem {...props} />
      </Provider>
    );

    expect(screen.getByLabelText(/test todo/i)).toBeInTheDocument();
    expect(screen.getByText(/test todo/i)).toBeInTheDocument();
  });

  it("should complete the todo when the checkbox is clicked", () => {
    render(
      <Provider store={store}>
        <TodoItem {...props} />
      </Provider>
    );

    fireEvent.click(screen.getByLabelText(/test todo/i));
    expect(store.getActions()).toContainEqual(completeTodo({ id: props.id }));
  });

  it("should delete the todo when the delete button is clicked", () => {
    render(
      <Provider store={store}>
        <TodoItem {...props} />
      </Provider>
    );

    fireEvent.click(screen.getByRole("button"));
    expect(store.getActions()).toContainEqual(deleteTodo({ id: props.id }));
  });

  it("should call the onDragStartHandle prop when the item is dragged", () => {
    render(
      <Provider store={store}>
        <TodoItem {...props} />
      </Provider>
    );

    fireEvent.dragStart(screen.getByRole("todo-item"));
    expect(props.onDragStartHandle).toHaveBeenCalled();
  });
});