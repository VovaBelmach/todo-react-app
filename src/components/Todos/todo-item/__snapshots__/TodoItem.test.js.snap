// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component should match snapshot 1`] = `
<div>
  <li
    class="incomplete"
    draggable="true"
    role="todo-item"
    style="opacity: 1;"
  >
    <label>
      <input
        class="checkbox"
        id="1"
        name="1"
        type="checkbox"
      />
      Test Todo
    </label>
    <button
      class="button delete-btn"
      type="button"
    >
      X
    </button>
  </li>
</div>
`;
